### ISSUE TITLE
Bazel fails with "Invalid RunRequest: bad cookie"

### ISSUE THREAD

**Issue by loshca** [bazelbuild/bazel#12190]
### Description of the problem / feature request:

Periodically Bazel fails with bad cookie error on our CI:
```
[INFO 16:11:39.607 src/main/cpp/blaze.cc:1256] Connected (server pid=2172).
[INFO 16:11:39.608 src/main/cpp/blaze.cc:1966] Releasing client lock, let the server manage concurrent requests.
[INFO 16:11:39.740 src/main/cpp/blaze.cc:2085] failure_detail: message: "Invalid RunRequest: bad cookie"
grpc_server {
    code: BAD_COOKIE
}
```

We have CI with two sequential launches of docker container with Bazel invocations in both and docker launches use a common mounted directory for the Bazel "outputRoot".

In other words: we have few Bazel launches then killing a Bazel server process (no proper shutdown, just stop the container) and then have other few launches.

Periodically we have a "bad cookie" error in the second part.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

It happens in less than 1% of builds, but in our case, it's every few days.
Basically, you have to launch Bazel, kill a server process, launch Bazel again, repeat until happens.

### What operating system are you running Bazel on?

windows/linux

### What's the output of `bazel info release`?

release 3.5.0

###  Have you found anything relevant by searching the web?

No.

#### My investigation
Bad cookie error coming from
https://github.com/bazelbuild/bazel/blob/4e11dc162f41e1d5ab5290d941064005c8347b9a/src/main/java/com/google/devtools/build/lib/server/GrpcServerImpl.java#L484
when request cookie from the client doesn't match the server one.

A server generates a request cookie as a random value on its creation:
https://github.com/bazelbuild/bazel/blob/4e11dc162f41e1d5ab5290d941064005c8347b9a/src/main/java/com/google/devtools/build/lib/server/GrpcServerImpl.java#L124

And writes it on disk in the server directory all together with the  port file and the response cookie file:
https://github.com/bazelbuild/bazel/blob/4e11dc162f41e1d5ab5290d941064005c8347b9a/src/main/java/com/google/devtools/build/lib/server/GrpcServerImpl.java#L439

Client waits until it can read all these files from the server directory in the [blaze.cc](https://github.com/bazelbuild/bazel/blob/master/src/main/cpp/blaze.cc) `BlazeServer::Connect()`.

When there is a first "clean" run with no sever directory at all or after `bazel shutdown` it works just fine: client waits until all files will be written and then connect to the server (or timeout will happen).

But if Bazel server process is killed or crashed then no cleanup actions happen, and all three files are left in the server directory. And we need some protection to keep the client from reading port and cookies from the old server, and there is such protection, but only for the port file. Client remove port file (if exists) before launch server daemon:
https://github.com/bazelbuild/bazel/blob/389e0da1fedc7543e243267a1052f9209b4dd395/src/main/cpp/blaze.cc#L863

But all these files written by the server sequentialy, not atomically:
```
    writeServerFile(PORT_FILE, addressString);
    writeServerFile(REQUEST_COOKIE_FILE, requestCookie);
    writeServerFile(RESPONSE_COOKIE_FILE, responseCookie);
```

So sometimes the client "wins" server and read the old cookie before sever rewrites it, which cause a "bad cookie" error.
I consider to remove cookie files with the port file in my PR #12189. Please check it out.

**Comment by lberki** [comment]
Wow. Nice diagnosis! I thought we handle this case properly, but I was apparently wrong.

**Comment by lberki** [comment]
I don't understand, though, how Bazel can fail. The "Connected" message comes from line 1256 and that means that the `StartServerAndConnect()` call in line 1224 worked out and that calls `ConnectOrDie()` (in line 889), which tries to call `server->Connect()` in a loop in line 791 and retries if it returns false.

But `server->Connect()` calls `TryConnect()` in turn in line 1753, which should return false if the cookie is wrong, shouldn' it?

**Comment by loshca** [comment]
Yes, you're right, I didn't notice the "TryConnect" call when I was dealing with the error for the first time.
And we caught this error again, after the patch - so my PR doesn't fix the error.

I have a theory: "writeServerFile" does not write the file atomically, this why "TryConnect" does not fail: [openStream creates an empty file before adding anything to it](https://github.com/google/guava/blob/0cd4e9faa1360da4a343f84cb275d6eda0c5e732/guava/src/com/google/common/io/CharSink.java#L97). As a result, the response cookie "matches" because the server sent an empty response cookie (the request cookie did not match), and the client read the empty cookie from the file since the server had not yet called a flush on it.

I patched our Bazel to make "writeServerFile" to write a file the same way as "serverInfoFile": write all content to the temp file, then atomically move it to the final destination.

I also logged cookies values read by the client and cookies generated by the server. If we will get the error again, I at least will give you some additional information.

**Comment by lberki** [comment]
That could be. If that theory is correct, you could fix it (at least temporarily) by making the server set the response cookie to some non-empty value when the request cookie does not match. 

Another, less hackish way would be simply to write the command port file after the cookie files. It would still be subject to the same race condition (but there it's handled in a better way), so the write-then-rename procedure still makes sense (you'd have to be a bit careful so that the case where the server dies after writing a file but before renaming it works well)

**Comment by michajlo** [comment]
Figuring out how to consolidate all interactions into `server_info.rawproto` could be a nice long-term fix, but I suspect it'll take more time and effort than you're interested in taking on (which is why I haven't done it yet).

**Comment by thekyz** [comment]
@loshca did you ever encounter the error after that workaround?

**Comment by loshca** [comment]
> @loshca did you ever encounter the error after that workaround?

@thekyz Nope, now works as a charm.
